File: core/src/io/github/jerukan/planetdata/PlanetState.java
package io.github.jerukan.planetdata;

import com.badlogic.gdx.math.Vector2;
import io.github.jerukan.physics.PhysicsObject;
import io.github.jerukan.physics.PhysicsState;
import io.github.jerukan.rendering.Drawable;

import java.util.ArrayList;

public class PlanetState extends PhysicsState {

    private ArrayList<Planet> planetList;
    private ArrayList<Drawable> drawables;

    public PlanetState(ArrayList<Planet> planetList, ArrayList<Drawable> drawables) {
        this.planetList = planetList;
        this.drawables = drawables;
    }

    @Override
    public void update() {
        for(int loops = 0; loops < warp; loops++) {
            for (int i = 0; i < planetList.size(); i++) {
                for (int j = i + 1; j < planetList.size(); j++) {
                    if (planetList.get(i).collidesPlanet(planetList.get(j))) {
                        planetList.get(i).onCollision(planetList.get(j));
                        planetList.get(j).onCollision(planetList.get(i));
                    }
                }
                planetList.get(i).updateVectors((float)timer.getDelta());
            }
            timer.update();
        }
    }

    public void add(Planet planet) {
        planetList.add(planet);
        planet.setGravitiesFromPlanets(planetList);
        drawables.add(planet);
    }

    public void add(Planet ... allPlanets) {
        for(Planet planet : allPlanets) {
            add(planet);
        }
    }

    @Override
    public void add(PhysicsObject physicsObject) {
        if(physicsObject instanceof Planet) {
            add((Planet) physicsObject);
        } else {
            super.add(physicsObject);
        }
    }
}

File: core/src/io/github/jerukan/planetdata/PlanetState.kt
package io.github.jerukan.planetdata

import com.badlogic.gdx.math.Vector2
import io.github.jerukan.physics.PhysicsObject
import io.github.jerukan.physics.PhysicsState
import io.github.jerukan.rendering.Drawable

class PlanetState(private val planetList: ArrayList<Planet>, private val drawables: ArrayList<Drawable>): PhysicsState() {
//    var mercury: Planet = Planet("mercury")
//    var venus: Planet = Planet("venus")
//    var earth: Planet = Planet("earth")
//    var mars: Planet = Planet("mars")
//    var jupiter: Planet = Planet("jupiter")
//    var saturn: Planet = Planet("saturn")
//    var uranus: Planet = Planet("uranus")
//    var neptune: Planet = Planet("neptune")
//    var pluto: Planet = Planet("pluto")
//
//    var planetList: Array<Planet> = arrayOf(mercury, venus, earth, mars, jupiter, saturn, uranus, neptune, pluto)
//    var p1 = Planet("p1", 500f, Vector2(-200f, 0f), 50f)
//    var p2 = Planet("p2", 500f, Vector2(200f, 0f), 50f)
//    var p3 = Planet("p3", 200f, Vector2(500f, 100f), 30f)

    override fun update() {
        for(loops in 0 until warp) {
            for (i in planetList.indices) {
                for (j in IntRange(i + 1, planetList.size - 1)) {
                    if (planetList[i].collidesPlanet(planetList[j])) {
                        planetList[i].onCollision(planetList[j])
                        planetList[j].onCollision(planetList[i])
                    }
                }
                planetList[i].updateVectors(timer.getDelta().toFloat())
            }
            timer.update()
        }
    }

    fun add(planet: Planet) {
        planetList.add(planet)
        planet.setGravitiesFromPlanets(planetList)
        drawables.add(planet)
    }

    fun add(vararg allPlanets: Planet) {
        for(planet in allPlanets) {
            add(planet)
        }
    }

    override fun add(physicsObject: PhysicsObject) {
        if(physicsObject is Planet) {
            add(physicsObject)
        } else {
            super.add(physicsObject)
        }
    }
}